containers/Connection.ts

```diff
@@ -1,54 +1,116 @@
 import { createContainer } from "unstated-next";
 import { useState, useEffect } from "react";
 import { ethers } from "ethers";
+import { API as OnboardApi, Wallet } from "bnc-onboard/dist/src/interfaces";
+import { API as NotifyApi } from "bnc-notify/dist/src/interfaces";
+import Onboard from "bnc-onboard";
+import Notify from "bnc-notify";
 import { Observable } from "rxjs";
 import { debounceTime } from "rxjs/operators";
 
 type Provider = ethers.providers.Provider;
 type Block = ethers.providers.Block;
 type Network = ethers.providers.Network;
-type ExternalProvider = ethers.providers.ExternalProvider;
 type Signer = ethers.Signer;
 
 function useConnection() {
   const [provider, setProvider] = useState<Provider | null>(null);
+  const [onboard, setOnboard] = useState<OnboardApi | null>(null);
+  const [notify, setNotify] = useState<NotifyApi | null>(null);
   const [signer, setSigner] = useState<Signer | null>(null);
   const [network, setNetwork] = useState<Network | null>(null);
   const [address, setAddress] = useState<string | null>(null);
+  const [selectedWallet, setSelectedWallet] = useState<string | null>(null);
   const [error, setError] = useState<Error | null>(null);
   const [block$, setBlock$] = useState<Observable<Block> | null>(null);
 
   const attemptConnection = async () => {
-    if (window.ethereum === undefined) {
-      throw Error("MetaMask not found, please visit https://metamask.io/");
-    }
+    const apiKey = process.env.REACT_APP_ONBOARD_API_KEY
```

Comment:
note the dapp will work fine without specifying any of these params. They default to common public values (public Infura key & the default onboard.js API key)

Reply:
Going off matt's questions, who's keys are these and why would you ever want to set your own keys?

Reply:
This is the sample key from the onboard.js repo. note that not all wallets need this param. Metamask, for example, does not require you to provide this key. 

I'm not actually sure how the dos protection would work in this case. I mean if it's set as a .env variable for the server then the risk is pretty low as no one can actually find it and dos it. In the official docs it's just that some wallet providers require this as a field 🤷‍♂️ https://docs.blocknative.com/onboard#built-in-modules
I can reach out to them and ask about this specific point as I'm not actually sure.

Reply:
Is there any way to avoid having these env vars in the dApp? They're not very clean and I'd like to avoid having these .env dependencies even if they're not a problem in run-time.

Reply:
well this is why it's optional and there are defaults :) if you don't provide them the app still works fine but ideally, you want to have them.

Onboard.js, if provided with your own key, is pretty useful. it gives metrics of the users that connect their wallets and statistics around them.

Reply:
I do agree though that this is the main downside with onboard.js. If we feel really strongly about this we can try a different provider, like [web3modal](https://github.com/Web3Modal/web3modal). I do think though that some providers will _always_ need some kind of keys, such as portis/fortmatic/ImToken.

Reply:
Ok I'm good with this and using our own onboard.js key since we can get metrics!

However, do you know how we set configs in Vercel?

Reply:
Great. It's pretty [easy](https://vercel.com/blog/environment-variables-ui#:~:text=New%20Project%20Setting,field%20in%20the%20Project%20Settings.) to do env vars in Vercel. I'll set them up now before merging this in.